### PR TITLE
fix when filter is specified, Activate annotation parameters need to be verified.

### PR DESCRIPTION
**Pull request by xiaoheng1**
## What is the purpose of the change

fix when filter is specified, Activate annotation parameters need to be verified.
see more detail from #7416 

## Brief changelog

XXXXX

## Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java [hunk 1]
```diff
@@ -260,6 +262,9 @@ public String getExtensionName(Class<?> extensionClass) {
     public List<T> getActivateExtension(URL url, String[] values, String group) {
         List<T> activateExtensions = new ArrayList<>();
         List<String> names = values == null ? new ArrayList<>(0) : asList(values);
+        // store not match filter.
+        Set<String> notMatchFilters = new HashSet<>();
```
Comment (horizonzy) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java:266]: just need notMatchFilters is enough.

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java [hunk 2]
```diff
@@ -297,6 +305,14 @@ public String getExtensionName(Class<?> extensionClass) {
                         loadedExtensions.clear();
                     }
                 } else {
+                    if (notMatchFilters.contains(name)) {
```
Comment (horizonzy) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java:308]: if a filter config dubbo.consumer.filter, the filter always be notMatch, whatever the group is match or not match.

File: dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java [hunk 3]
```diff
@@ -297,6 +305,14 @@ public String getExtensionName(Class<?> extensionClass) {
...
+                        logger.warn(name + " filter that does not meet the Activate condition has been loaded.");
+                    }
+                    if (matchFilters.contains(name)) {
```
Comment (horizonzy) [inline dubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java:311]: the verify just log info?

**Comment by islandValent** [comment]
unsubscribe



At 2021-03-20 17:33:42, "xiaoheng1" ***@***.***> wrote:

What is the purpose of the change

fix when filter is specified, Activate annotation parameters need to be verified.
see more detail from #7416

Brief changelog

XXXXX

Verifying this change

XXXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

Make sure there is a GITHUB_issue field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
Format the pull request title like [Dubbo-XXX] Fix UnknownException when host config not exist #XXX. Each commit in the pull request should have a meaningful subject line and body.
Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in dubbo samples project.
Run mvn clean install -DskipTests=false & mvn clean test-compile failsafe:integration-test to make sure unit-test and integration-test pass.
If this contribution is large, please follow the Software Donation Guide.
You can view, comment on, or merge this pull request online at:

  https://github.com/apache/dubbo/pull/7417

Commit Summary
fix #7416 when filter is specified, Activate annotation parameters need to be verified.
File Changes
Mdubbo-common/src/main/java/org/apache/dubbo/common/extension/ExtensionLoader.java (36)
Patch Links:
https://github.com/apache/dubbo/pull/7417.patch
https://github.com/apache/dubbo/pull/7417.diff

**Comment by chickenlj** [comment]
I think there's no need to add this check.
* Activate should not restrict the user from activating one extension or not.
* Keep ExtensionLoader simple, leave the constriction to users.

**Comment by xiaoheng1** [comment]
> I think there's no need to add this check.
> 
> * Activate should not restrict the user from activating one extension or not.
> * Keep ExtensionLoader simple, leave the constriction to users.

OK.
I  will turn off this pr.

**Comment by chickenlj** [comment]
Thanks.

**Pull request closed without merging**